plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -1678,15 +1697,15 @@ private void flushBuffer(boolean finished)
             }
 
             // The multipart upload API only accept the last part to be less than 5MB
-            if (bufferSize == buffer.length || (finished && bufferSize > 0)) {
+            if (bufferSize == partSize || (finished && bufferSize > 0)) {
                 byte[] data = buffer;
                 int length = bufferSize;
 
                 if (finished) {
                     this.buffer = null;
                 }
                 else {
-                    this.buffer = new byte[buffer.length];
+                    this.buffer = new byte[partSize];
```

Comment:
this is a noop (`buffer.length == buffer.length`).

What I mean is to use:
```
this.buffer = new byte[0];
```
here and then in subsequent `ensureExtraBytesCapacity` use `newBytesLength = partSize`.
You can make `INITIAL_BUFFER_SIZE` a `field`, like `initialBufferSize` and
* set it to `64` in constructor
* set it to `partSize` here, e.g:
```
this.buffer = new byte[0];
this.initialBufferSize = partSize
```

Reply:
So actually this won't work as it is. 
Either we have to do
```
this.buffer = new byte[partSize];
``` 
here or in the `public void write(byte[] bytes, int offset, int length)` method call  `ensureExtraBytesCapacity` inside the loop. Otherwise it will hang because no progress will be made in the while loop.
I opt towards first option as it is more performant.

Reply:
Oh, I see, but:
```
this.initialBufferSize = partSize;
```
is really noop here now. You can do `ensureExtraBytesCapacity` in loop as in https://github.com/trinodb/trino/pull/14212/commits/8408da0af9aaf29df359e1d719a3bd6ea1970343#r986804499